### PR TITLE
async_udp.v0.12.0

### PR DESCRIPTION
**Pull request by Leonidas-from-XIV**
Please note that this package is meant to be replaced once the official async_udp package from Jane Street is released.

### CODE REVIEW

**Comment by kit-ty-kate** [comment]
cc @xclerc ?

**Comment by xclerc** [comment]
There is indeed an issue with UDP, which is now part
of [async_extra](https://github.com/janestreet/async_extra) (in master, not in the latest release).

As we expect to do a stable release (v0.13.0) of our
packages soon, we might want to avoid this intermediary
step of having a standalone `async_udp` package.

**Comment by Leonidas-from-XIV** [comment]
The problem is that Udp has been missing ever since v0.12.0 was released ([ticket](https://github.com/janestreet/async_extra/issues/13)), and thus preventing everybody who uses async_udp from upgrading OCaml 4.07, since base v0.11 (a transitive dependency of async v0.11) is constrained to < OCaml 4.08.

**Comment by xclerc** [comment]
One problem, I think is that in v0.12, `async` depends
on `async_extra`. It is no longer the case in master (and
thus upcoming v0.13).

If there is a standalone release of `async_udp`, could we
at least make it obvious from its name that it is a kind
of "transition" package.

(Also, I would like to have a second opinion on this PR,
thus cc @trefis.)

**Comment by trefis** [comment]
What I don't understand is why we don't just release `async_extra.v0.12.X` which includes async_udp. I think that was the plan at some point, but that didn't happen.
Do you know why?

**Comment by xclerc** [comment]
The problem is that `async_udp` (which is part of `async_extra`)
depends on `async`, which `include`s `async_extra`.

My understanding is that for the point release to happen, we would
need to remove the aforementioned `include`, thus breaking
compatibility in a point release. (I would also have to double check
for possible further consequences of the removal.)

**Comment by trefis** [comment]
Thanks for the explanation!

I had a discussion offline with @xclerc, here are the highlights:
- we should have communicated the problem on https://github.com/janestreet/async_extra/issues/13
- `async_udp` in our repository is part of the `async_extra` package: that is because we didn't want to create opam packages too aggressively
- that is preventing us from releasing a v0.12: because of the cyclic-dependency problem described above; the problem will not exist in v0.13
- this PR is creating an `async_udp` package
- if this PR is merged, but the next version (v0.13) of `async_udp` is part of the `async_extra` package that's likely to be confusing for user (the package will seem to have disappeared)
- to avoid this confusion, we have decided to just split out `async_udp` into its own package
- since we're splitting `async_udp` out, we might as well release a v0.12 ourselves

tl;dr: We're going to create an `async_udp` repository and package, and we're going to make an opam released on v0.12. All of this should happen over the next few days.

Therefore, my opinion is that this PR should be closed without merging. (Because the ownership story is clearer if we do the release ourselves).

@kit-ty-kate: does that all make sense to you?

**Comment by avsm** [comment]
Thanks for the detailed explanation @trefis. Regarding:

> Therefore, my opinion is that this PR should be closed without merging. (Because the ownership story is clearer if we do the release ourselves).

My strong preference is for ownership to remain with JS for all of these packages if possible, and to not create an exception for this sole release of async_udp. There are quite a few automated scripts that generate these Jane Street packages to opam-repository, so let's avoid exceptions if possible.

**Comment by Leonidas-from-XIV** [comment]
@trefis Thanks for the summary. This sounds like the optimal outcome to me; I also would prefer for JS to keep ownership of it. My rationale about this PR was to make it possible to have v0.12 and UDP, but if a JS-owned `async_udp` package exists I would indeed prefer that.
